File: data_structures/ruby/graph/directed_acyclic_graph.rb
require 'set'

class DAG
  def initialize
    # Adjacency list to store the graph (task -> list of dependent tasks)
    @graph = Hash.new { |h, k| h[k] = [] }

    # Store task descriptions
    @tasks = {}
  end

  def add_task(task_id, description)
    @tasks[task_id] = description
  end

  def add_dependency(from_task, to_task)
    unless @tasks.key?(from_task) && @tasks.key?(to_task)
      raise "Both tasks must exist in the DAG."
    end

    @graph[from_task] << to_task
  end

  def has_cycle?
    visited = Set.new
    rec_stack = Set.new

    dfs = lambda do |task|
      visited.add(task)
      rec_stack.add(task)

      @graph[task].each do |neighbor|
        if !visited.include?(neighbor)
          return true if dfs.call(neighbor)
        elsif rec_stack.include?(neighbor)
          return true
        end
      end

      rec_stack.delete(task)
      false
    end

    @tasks.keys.each do |task|
      next if visited.include?(task)

      return true if dfs.call(task)
    end

    false
  end

  def topological_sort
    raise "DAG contains a cycle, cannot perform topological sort." if has_cycle?

    # Compute in-degree for each task
    in_degree = @tasks.keys.to_h { |task| [task, 0] }
    @graph.each do |_task, neighbors|
      neighbors.each do |neighbor|
        in_degree[neighbor] += 1
      end
    end

    # Initialize queue with tasks having in-degree 0
    queue = @tasks.keys.select { |task| in_degree[task] == 0 }
    result = []

    until queue.empty?
      current = queue.shift
      result << current

      # Reduce in-degree for neighbors
      @graph[current].each do |neighbor|
        in_degree[neighbor] -= 1
        queue << neighbor if in_degree[neighbor] == 0
      end
    end

    if result.length != @tasks.length
      raise "DAG contains a cycle or disconnected components."
    end

    result
  end

  def print_tasks(task_order)
    puts "Task Execution Order:"
    task_order.each do |task|
      puts "#{task}: #{@tasks[task]}"
    end
  end

  attr_reader :graph, :tasks
end

def check_and_print_order(dag)
  if !dag.has_cycle?
    puts "\n\nNo cycles detected in the DAG.\n\n"
    task_order = dag.topological_sort
    dag.print_tasks(task_order)

    # Build dependency summary
    dependencies = []
    dag.graph.each do |task, neighbors|
      neighbors.each do |neighbor|
        dependencies << "#{task} (#{dag.tasks[task]}) -> #{neighbor} (#{dag.tasks[neighbor]})"
      end
    end

    # Print dependency summary
    puts "\nDependency Summary:"
    dependencies.sort.each { |dep| puts dep }
  else
    puts "\n\nCycle detected in the DAG.\n\n"
  end
end

def main
  dag = DAG.new

  # Add tasks
  dag.add_task("A", "Research topic")
  dag.add_task("B", "Create slides")
  dag.add_task("C", "Practice delivery")
  dag.add_task("D", "Finalize presentation")

  # Add dependencies
  dag.add_dependency("A", "B")  # B depends on A
  dag.add_dependency("B", "C")  # C depends on B
  dag.add_dependency("B", "D")  # D depends on B
  dag.add_dependency("C", "D")  # D depends on C

  # Check for cycles and print order
  check_and_print_order(dag)

  # Create a more complex DAG for software development lifecycle
  dag = DAG.new

  # Add tasks for different phases
  dag.add_task("REQ", "Requirements gathering")
  dag.add_task("SPEC", "Write technical specification")
  dag.add_task("DESIGN", "System architecture design")
  dag.add_task("DB", "Database schema design")
  dag.add_task("API", "API design")
  dag.add_task("UI", "UI/UX design")
  dag.add_task("CODE_BE", "Backend implementation")
  dag.add_task("CODE_FE", "Frontend implementation")
  dag.add_task("TEST_UNIT", "Unit testing")
  dag.add_task("TEST_INT", "Integration testing")
  dag.add_task("TEST_E2E", "End-to-end testing")
  dag.add_task("DOC", "Documentation")
  dag.add_task("REVIEW", "Code review")
  dag.add_task("DEPLOY", "Deployment")

  # Add dependencies
  dag.add_dependency("REQ", "SPEC")
  dag.add_dependency("SPEC", "DESIGN")
  dag.add_dependency("DESIGN", "DB")
  dag.add_dependency("DESIGN", "API")
  dag.add_dependency("DESIGN", "UI")
  dag.add_dependency("DB", "CODE_BE")
  dag.add_dependency("API", "CODE_BE")
  dag.add_dependency("API", "CODE_FE")
  dag.add_dependency("UI", "CODE_FE")
  dag.add_dependency("CODE_BE", "TEST_UNIT")
  dag.add_dependency("CODE_FE", "TEST_UNIT")
  dag.add_dependency("TEST_UNIT", "TEST_INT")
  dag.add_dependency("TEST_INT", "TEST_E2E")
  dag.add_dependency("CODE_BE", "DOC")
  dag.add_dependency("CODE_FE", "DOC")
  dag.add_dependency("TEST_UNIT", "REVIEW")
  dag.add_dependency("DOC", "REVIEW")
  dag.add_dependency("REVIEW", "DEPLOY")
  dag.add_dependency("TEST_E2E", "DEPLOY")

  # Check for cycles and print order
  check_and_print_order(dag)

  # Adding additional dependencies
  dag.add_dependency("REQ", "DB")       # DB now depends on both REQ and DESIGN
  dag.add_dependency("SPEC", "API")     # API now depends on both SPEC and DESIGN
  dag.add_dependency("TEST_INT", "DOC") # DOC now depends on CODE_BE, CODE_FE and TEST_INT
  dag.add_dependency("TEST_INT", "REVIEW") # REVIEW now depends on TEST_UNIT, DOC and TEST_INT

  # Check for cycles and print order
  check_and_print_order(dag)
end

if __FILE__ == $PROGRAM_NAME
  main
end

File: data_structures/go/graph/directed_acyclic_graph.go
package main

import (
	"errors"
	"fmt"
	"sort"
)

type DAG struct {
	graph map[string][]string
	tasks map[string]string
}

func NewDAG() *DAG {
	return &DAG{
		graph: make(map[string][]string),
		tasks: make(map[string]string),
	}
}

func (d *DAG) AddTask(taskID, description string) {
	d.tasks[taskID] = description
	if _, exists := d.graph[taskID]; !exists {
		d.graph[taskID] = []string{}
	}
}

func (d *DAG) AddDependency(fromTask, toTask string) error {
	if _, exists := d.tasks[fromTask]; !exists {
		return errors.New("from task does not exist in the DAG")
	}
	if _, exists := d.tasks[toTask]; !exists {
		return errors.New("to task does not exist in the DAG")
	}
	d.graph[fromTask] = append(d.graph[fromTask], toTask)
	return nil
}

func (d *DAG) HasCycle() bool {
	visited := make(map[string]bool)
	recStack := make(map[string]bool)

	var dfs func(task string) bool
	dfs = func(task string) bool {
		visited[task] = true
		recStack[task] = true

		for _, neighbor := range d.graph[task] {
			if !visited[neighbor] {
				if dfs(neighbor) {
					return true
				}
			} else if recStack[neighbor] {
				return true
			}
		}

		recStack[task] = false
		return false
	}

	for task := range d.tasks {
		if !visited[task] {
			if dfs(task) {
				return true
			}
		}
	}
	return false
}

func (d *DAG) TopologicalSort() ([]string, error) {
	if d.HasCycle() {
		return nil, errors.New("DAG contains a cycle, cannot perform topological sort")
	}

	// Compute in-degree for each task
	inDegree := make(map[string]int)
	for task := range d.tasks {
		inDegree[task] = 0
	}
	for _, neighbors := range d.graph {
		for _, neighbor := range neighbors {
			inDegree[neighbor]++
		}
	}

	// Initialize queue with tasks having in-degree 0
	var queue []string
	for task := range d.tasks {
		if inDegree[task] == 0 {
			queue = append(queue, task)
		}
	}
	sort.Strings(queue) // Sort for deterministic output

	var result []string

	for len(queue) > 0 {
		// Pop from front
		current := queue[0]
		queue = queue[1:]
		result = append(result, current)

		// Get neighbors and sort for deterministic order
		neighbors := make([]string, len(d.graph[current]))
		copy(neighbors, d.graph[current])
		sort.Strings(neighbors)

		// Reduce in-degree for neighbors
		for _, neighbor := range neighbors {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
				sort.Strings(queue)
			}
		}
	}

	if len(result) != len(d.tasks) {
		return nil, errors.New("DAG contains a cycle or disconnected components")
	}
	return result, nil
}

func (d *DAG) PrintTasks(taskOrder []string) {
	fmt.Println("Task Execution Order:")
	for _, task := range taskOrder {
		fmt.Printf("%s: %s\n", task, d.tasks[task])
	}
}

func checkAndPrintOrder(dag *DAG) {
	if !dag.HasCycle() {
		fmt.Println("\nNo cycles detected in the DAG.\n")
		taskOrder, err := dag.TopologicalSort()
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		dag.PrintTasks(taskOrder)

		// Build dependency summary
		var dependencies []string
		for task, neighbors := range dag.graph {
			for _, neighbor := range neighbors {
				dep := fmt.Sprintf("%s (%s) -> %s (%s)", task, dag.tasks[task], neighbor, dag.tasks[neighbor])
				dependencies = append(dependencies, dep)
			}
		}

		// Print dependency summary
		sort.Strings(dependencies)
		fmt.Println("\nDependency Summary:")
		for _, dep := range dependencies {
			fmt.Println(dep)
		}
	} else {
		fmt.Println("\nCycle detected in the DAG.\n")
	}
}

func main() {
	dag := NewDAG()

	// Add tasks
	dag.AddTask("A", "Research topic")
	dag.AddTask("B", "Create slides")
	dag.AddTask("C", "Practice delivery")
	dag.AddTask("D", "Finalize presentation")

	// Add dependencies
	dag.AddDependency("A", "B") // B depends on A
	dag.AddDependency("B", "C") // C depends on B
	dag.AddDependency("B", "D") // D depends on B
	dag.AddDependency("C", "D") // D depends on C

	// Check for cycles and print order
	checkAndPrintOrder(dag)

	// Create a more complex DAG for software development lifecycle
	dag = NewDAG()

	// Add tasks for different phases
	dag.AddTask("REQ", "Requirements gathering")
	dag.AddTask("SPEC", "Write technical specification")
	dag.AddTask("DESIGN", "System architecture design")
	dag.AddTask("DB", "Database schema design")
	dag.AddTask("API", "API design")
	dag.AddTask("UI", "UI/UX design")
	dag.AddTask("CODE_BE", "Backend implementation")
	dag.AddTask("CODE_FE", "Frontend implementation")
	dag.AddTask("TEST_UNIT", "Unit testing")
	dag.AddTask("TEST_INT", "Integration testing")
	dag.AddTask("TEST_E2E", "End-to-end testing")
	dag.AddTask("DOC", "Documentation")
	dag.AddTask("REVIEW", "Code review")
	dag.AddTask("DEPLOY", "Deployment")

	// Add dependencies
	dag.AddDependency("REQ", "SPEC")
	dag.AddDependency("SPEC", "DESIGN")
	dag.AddDependency("DESIGN", "DB")
	dag.AddDependency("DESIGN", "API")
	dag.AddDependency("DESIGN", "UI")
	dag.AddDependency("DB", "CODE_BE")
	dag.AddDependency("API", "CODE_BE")
	dag.AddDependency("API", "CODE_FE")
	dag.AddDependency("UI", "CODE_FE")
	dag.AddDependency("CODE_BE", "TEST_UNIT")
	dag.AddDependency("CODE_FE", "TEST_UNIT")
	dag.AddDependency("TEST_UNIT", "TEST_INT")
	dag.AddDependency("TEST_INT", "TEST_E2E")
	dag.AddDependency("CODE_BE", "DOC")
	dag.AddDependency("CODE_FE", "DOC")
	dag.AddDependency("TEST_UNIT", "REVIEW")
	dag.AddDependency("DOC", "REVIEW")
	dag.AddDependency("REVIEW", "DEPLOY")
	dag.AddDependency("TEST_E2E", "DEPLOY")

	// Check for cycles and print order
	checkAndPrintOrder(dag)

	// Adding additional dependencies
	dag.AddDependency("REQ", "DB")       // DB now depends on both REQ and DESIGN
	dag.AddDependency("SPEC", "API")     // API now depends on both SPEC and DESIGN
	dag.AddDependency("TEST_INT", "DOC") // DOC now depends on CODE_BE, CODE_FE and TEST_INT
	dag.AddDependency("TEST_INT", "REVIEW")

	// Check for cycles and print order
	checkAndPrintOrder(dag)
}
